### ISSUE TITLE
Docs: rewrite examples with `var` using `const` and `let`

### ISSUE THREAD

**Issue by rakleed** [eslint/eslint#19240]
### Docs page(s)

https://github.com/search?q=repo%3Aeslint%2Feslint+path%3A%2F%5Edocs%5C%2Fsrc%5C%2Frules%5C%2F%2F+NOT+path%3A**%2F*var*+%22var+%22+&type=code

### What documentation issue do you want to solve?

Many examples use `var`, which is [considered bad practice](https://stackoverflow.com/questions/65663361/why-many-examples-of-javascript-still-using-var-instead-of-const-and-let#:~:text=So%20should%20you,style%20and%20change.) as it often [leads to errors](https://stackoverflow.com/questions/762011/what-is-the-difference-between-let-and-var#:~:text=The%20reason%20why%20let%20keyword%20was%20introduced%20to%20the%20language%20was%20function%20scope%20is%20confusing%20and%20was%20one%20of%20the%20main%20sources%20of%20bugs%20in%20JavaScript.). `const` and `let` have had good support in all browsers [for a long time](https://compat-table.github.io/compat-table/es6/).

### What do you think is the correct solution?

Rewrite examples that use `var` using `const` and `let`.

### Participation

- [ ] I am willing to submit a pull request for this change.

**Comment by nzakas** [comment]
No objection, we'd just need to be careful that the examples don't require `var` to be accurate. In some cases, like `no-var`, we'd obviously want `var` in the examples.

We'd love it if someone from the community would take this on as the core team has limited time. We don't have to do all of the pages at once, even PRs for one file at a time are welcome.

@rakleed do you want to submit a PR?

**Comment by amareshsm** [comment]
> No objection, we'd just need to be careful that the examples don't require `var` to be accurate. In some cases, like `no-var`, we'd obviously want `var` in the examples.
> 
> We'd love it if someone from the community would take this on as the core team has limited time. We don't have to do all of the pages at once, even PRs for one file at a time are welcome.
> 
> [@rakleed](https://github.com/rakleed) do you want to submit a PR?

I will make the changes for some of the rules @rakleed. If you're interested, feel free to pick any of them and submit a PR.

**Comment by amareshsm** [comment]
Should we do this update to the deprecated rules as well?

**Comment by nzakas** [comment]
I think it's fine to leave deprecated and removed rules alone.

**Comment by kr-vaibhav** [comment]
@nzakas @rakleed 
I'd like to work on this issue and make the required changes.
Please let me know if you'd like me to proceed with this.

**Comment by nzakas** [comment]
@kr-vaibhav thanks for volunteering. I think the best way to start is to review the existing rule documentation and create a list of the pages that need updating. You can post that list onto this issue as a checkbox list.

Then, we can have people take five at a time to convert in one PR. That way, we can keep the process moving without getting bogged down on large PRs.

How does that sound?

**Comment by amareshsm** [comment]
> [@kr-vaibhav](https://github.com/kr-vaibhav) thanks for volunteering. I think the best way to start is to review the existing rule documentation and create a list of the pages that need updating. You can post that list onto this issue as a checkbox list.
> 
> Then, we can have people take five at a time to convert in one PR. That way, we can keep the process moving without getting bogged down on large PRs.
> 
> How does that sound?

That sounds good. Initially, I raised a PR with changes to 8 rule files, but in the upcoming PR, we can limit it to 5.

**Comment by amareshsm** [comment]
Here's the list of all non-deprecated rules.



| Name | Status |
|------|--------|
| array-callback-return |        |
| constructor-super |        |
| for-direction |        |
| getter-return |        |
| no-async-promise-executor |        |
| no-await-in-loop |        |
| no-class-assign |        |
| no-compare-neg-zero |        |
| no-cond-assign |        |
| no-const-assign |        |
| no-constant-binary-expression |        |
| no-constant-condition |        |
| no-constructor-return |        |
| no-control-regex |        |
| no-debugger |        |
| no-dupe-args |        |
| no-dupe-class-members |        |
| no-dupe-else-if |        |
| no-dupe-keys |        |
| no-duplicate-case |        |
| no-duplicate-imports |        |
| no-empty-character-class |        |
| no-empty-pattern |        |
| no-ex-assign |        |
| no-fallthrough |        |
| no-func-assign |        |
| no-import-assign |        |
| no-inner-declarations |        |
| no-invalid-regexp |        |
| no-irregular-whitespace |        |
| no-loss-of-precision |        |
| no-misleading-character-class |        |
| no-new-native-nonconstructor |        |
| no-obj-calls |        |
| no-promise-executor-return |        |
| no-prototype-builtins |        |
| no-self-assign |        |
| no-self-compare |        |
| no-setter-return |        |
| no-sparse-arrays |        |
| no-template-curly-in-string |        |
| no-this-before-super |        |
| no-undef |        |
| no-unexpected-multiline |        |
| no-unmodified-loop-condition |        |
| no-unreachable |        |
| no-unreachable-loop |        |
| no-unsafe-finally |        |
| no-unsafe-negation |        |
| no-unsafe-optional-chaining |        |
| no-unused-private-class-members |        |
| no-unused-vars |        |
| no-use-before-define |        |
| no-useless-assignment |        |
| no-useless-backreference |        |
| require-atomic-updates |        |
| use-isnan |        |
| valid-typeof |        |
| accessor-pairs |        |
| arrow-body-style |        |
| block-scoped-var |        |
| camelcase |        |
| capitalized-comments |        |
| class-methods-use-this |        |
| complexity |        |
| consistent-return |        |
| consistent-this |        |
| curly |        |
| default-case |        |
| default-case-last |        |
| default-param-last |        |
| dot-notation |        |
| eqeqeq |        |
| func-name-matching |        |
| func-names |        |
| func-style |        |
| grouped-accessor-pairs |        |
| guard-for-in |        |
| id-denylist |        |
| id-length |        |
| id-match |        |
| init-declarations |        |
| logical-assignment-operators |        |
| max-classes-per-file |        |
| max-depth |        |
| max-lines |        |
| max-lines-per-function |        |
| max-nested-callbacks |        |
| max-params |        |
| max-statements |        |
| new-cap |        |
| no-alert |        |
| no-array-constructor |        |
| no-bitwise |        |
| no-caller |        |
| no-case-declarations |        |
| no-console |        |
| no-continue |        |
| no-delete-var |        |
| no-div-regex |        |
| no-else-return |        |
| no-empty |        |
| no-empty-function |        |
| no-empty-static-block |        |
| no-eq-null |        |
| no-eval |        |
| no-extend-native |        |
| no-extra-bind |        |
| no-extra-boolean-cast |        |
| no-extra-label |        |
| no-global-assign |        |
| no-implicit-coercion |        |
| no-implicit-globals |        |
| no-implied-eval |        |
| no-inline-comments |        |
| no-invalid-this |        |
| no-iterator |        |
| no-label-var |        |
| no-labels |        |
| no-lone-blocks |        |
| no-lonely-if |        |
| no-loop-func |        |
| no-magic-numbers |        |
| no-multi-assign |        |
| no-multi-str |        |
| no-negated-condition |        |
| no-nested-ternary |        |
| no-new |        |
| no-new-func |        |
| no-new-wrappers |        |
| no-nonoctal-decimal-escape |        |
| no-object-constructor |        |
| no-octal |        |
| no-octal-escape |        |
| no-param-reassign |        |
| no-plusplus |        |
| no-proto |        |
| no-redeclare |        |
| no-regex-spaces |        |
| no-restricted-exports |        |
| no-restricted-globals |        |
| no-restricted-imports |        |
| no-restricted-properties |        |
| no-restricted-syntax |        |
| no-return-assign |        |
| no-script-url |        |
| no-sequences |        |
| no-shadow |        |
| no-shadow-restricted-names |        |
| no-ternary |        |
| no-throw-literal |        |
| no-undef-init |        |
| no-undefined |        |
| no-underscore-dangle |        |
| no-unneeded-ternary |        |
| no-unused-expressions |        |
| no-unused-labels |        |
| no-useless-call |        |
| no-useless-catch |        |
| no-useless-computed-key |        |
| no-useless-concat |        |
| no-useless-constructor |        |
| no-useless-escape |        |
| no-useless-rename |        |
| no-useless-return |        |
| no-var |        |
| no-void |        |
| no-warning-comments |        |
| no-with |        |
| object-shorthand |        |
| one-var |        |
| operator-assignment |        |
| prefer-arrow-callback |        |
| prefer-const |        |
| prefer-destructuring |        |
| prefer-exponentiation-operator |        |
| prefer-named-capture-group |        |
| prefer-numeric-literals |        |
| prefer-object-has-own |        |
| prefer-object-spread |        |
| prefer-promise-reject-errors |        |
| prefer-regex-literals |        |
| prefer-rest-params |        |
| prefer-spread |        |
| prefer-template |        |
| radix |        |
| require-await |        |
| require-unicode-regexp |        |
| require-yield |        |
| sort-imports |        |
| sort-keys |        |
| sort-vars |        |
| strict |        |
| symbol-description |        |
| vars-on-top |        |
| yoda |        |
| unicode-bom |        |

**Comment by RichardLitt** [comment]
- `strict.js` appears to be done already.
- I have attempted to change all of the `var` in tests/lib/rules.js, with bad results - 16 of them fail, but I can't easily figure out which ones. There are no line numbers, for instance. Should we perhaps add those to make it easier?

**Comment by Tanujkanti4441** [comment]
> * `strict.js` appears to be done already.

strict has a `var`

https://github.com/eslint/eslint/blob/a559009f51ad9f081bae5252bb2b7a6e23c54767/docs/src/rules/strict.md?plain=1#L213

>I have attempted to change all of the var in tests/lib/rules.js, with bad results - 16 of them fail, but I can't easily figure out which ones. There are no line numbers, for instance. Should we perhaps add those to make it easier?

 Did you change `var` for all the rules? if yes then as explained above for some rules example `var` is required.
>we'd just need to be careful that the examples don't require var to be accurate. In some cases, like no-var, we'd obviously want var in the examples.

**Comment by ruddnjs3769** [comment]
Hello! I have reviewed all the rules where `no-useless` is used, and I found that there are no instances of `var` to modify in the files related to the following rules:

| Rule Name                 | Status   |
| ------------------------- | -------- |
| no-useless-call           | no var   |
| no-useless-catch          | no var   |
| no-useless-computed-key   |          |
| no-useless-concat         |          |
| no-useless-constructor    | no var   |
| no-useless-escape         | no var   |
| no-useless-rename         | no var   |
| no-useless-return         |          |

Therefore, there’s no need to modify `var`, and it looks like I only need to update three files related to `no-useless-call`, `no-useless-catch`, and `no-useless-computed-key`. I haven’t made the changes yet, but may I proceed with updating these three files and submitting a PR?

Thank you! :)

**Comment by RichardLitt** [comment]
> Did you change var for all the rules? if yes then as explained above for some rules example var is required.

Yes, as I mentioned, I tried to change them all. However, as mentioned: 

> 16 of them fail, but I can't easily figure out which ones. There are no line numbers, for instance. Should we perhaps add those to make it easier?

I can accept that I am not learned enough to understand the problem space, but maybe it would be worth explaining how I could learn about it. 

@ruddnjs3769: "I haven’t made the changes yet, but may I proceed with updating these three files and submitting a PR?"

For a small PR like this, I think you'd be fine making a PR. It's easier to do than to wait on someone else to ask permission, and it could always be closed or edited later.

**Comment by amareshsm** [comment]
@RichardLitt  Would you like to do changes to a few more rules?

**Comment by RichardLitt** [comment]
Traveling this week. Will see if I can on the other side. :)

**Comment by mdjermanovic** [comment]
Mistakenly closed by https://github.com/eslint/eslint/pull/19325.

**Comment by dharbrueger** [comment]
@Tanujkanti4441 I'd like to work on this a bit more today and maybe finish up the rest. Is the list above updated and mainly what we're targeting in this issue?

**Comment by RichardLitt** [comment]
I have run into more time constraints and am going to back away from helping for now. Wish I had a million of me. :)

**Comment by nzakas** [comment]
Thanks @RichardLitt :pray: We really appreciate it.

**Comment by snitin315** [comment]
All the rules mentioned in https://github.com/eslint/eslint/issues/19240#issuecomment-2566684802 have been updated to use `const`, I think we can close this issue now.

**Comment by aditya-hansraj** [comment]
@rakleed Hey, this issue is still open!
is any more help needed?
I am willing to help.
or you can just close the issue if completed!

**Comment by snitin315** [comment]
I just checked the [search results](https://github.com/search?q=repo%3Aeslint%2Feslint+path%3A%2F%5Edocs%5C%2Fsrc%5C%2Frules%5C%2F%2F+NOT+path%3A**%2F*var*+%22var+%22+&type=code), and they only contain examples from deprecated rules or cases where var usage is valid and can't be migrated to const.

The only remaining example to update is in `no-else-return`:

```js
function foo2() {
    if (x) {
        return y;
    } else if (z) {
        var t = "foo";
    } else {
        return w;
    }
}

function foo3() {
    if (x) {
        return y;
    } else {
        var t = "foo";
    }

    return t;
}

```

Once this is updated, we can close the issue.

**Comment by snitin315** [comment]
All examples have been migrated. Thank you to everyone who contributed to this issue! Closing it now. Feel free to submit a PR if you encounter any pending examples.

**Comment by Shreyas-Sarkar** [comment]
@rakleed @nzakas Hey, this issue is still open!
is any more help needed?
I am willing to help.
